### PR TITLE
DataLinks: Handle getLinks() regen during data updates and frame joining

### LINKED ISSUE #83573: Dashboards : pre config. Data link shows wrong TraceID in tooltip mode and leads you to a different trace
### What happened?

I've identified a potential issue with the traces (Tempo) where you can visualize them using for example a bar chart. Clicking on a trace reveals details such as trace duration or the trace ID. At the bottom of the list, you'll find a blue link labeled 'TRACE: (traceID).' This link directs you to a trace panel where you can access more information about the specific link. However, there seems to be a problem – the blue link displays the wrong trace ID. If the bar chart shows only one trace, the trace ID is correct. However, when there are two or more traces, all trace IDs appear to be switched.

Tarce issue1
Trace issue 2

### What did you expect to happen?

The TraceID link should be in every Bar correct/similair to the TraceID in the tooltips

### Did this work before?

I used it for the first time so Iam not sure. But for me i didnt work before.

### How do we reproduce it?

1. Create a new visualation and use the builder mode (Tempo) and create a bar chart or every other panel type except table, logs or trace.
2. Activate the tooltip mode and make sure the panel is showing more then one trace.
3. Click on one trace and compare the Trace ID and the link below You should see the issue

### Is the bug inside a dashboard panel?

Yes i use as panel a bar chart. But the issue is nearly on every other panel type (except table) where you can activate the tooltip mode

### Environment (with versions)?

Grafana: v10.3.3
OS: Linux
Browser: Firefox

### Grafana platform?

Kubernetes

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by leeoniya**
this removes or refactors various bits from a host of previous PRs. it puts things into the correct places, and fixes a bunch of bugs and performance footguns.

- https://github.com/grafana/grafana/pull/71811
- https://github.com/grafana/grafana/pull/60011
- https://github.com/grafana/grafana/pull/58424
- https://github.com/grafana/grafana/pull/56729

essentially this moves the `field.getLinks` regen out of the render path and into the data prep / frame joining path. this allows it to work correctly with both old and new tooltips. additionally, it fixes the dataLink case when multiple frames are mis-aligned (see dashboard below) and likely other hidden issues with incorrect frame length, etc.

maybe it fixes https://github.com/grafana/grafana/issues/83573 :shrug: 

this is a modified version of the repro dashboard outlined in https://github.com/grafana/grafana/issues/58307#issue-1437948659

<details><summary>data-links.json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 970,
  "links": [],
  "panels": [
    {
      "datasource": {
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "links": [
            {
              "title": "number ${__data.fields.numberField} string ${__data.fields.stringField}",
              "url": "number ${__data.fields.numberField} string ${__data.fields.stringField}"
            }
          ],
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "numberField2"
            },
            "properties": [
              {
                "id": "links",
                "value": [
                  {
                    "title": "number ${__data.fields.numberField2} string ${__data.fields.stringField2}",
                    "url": "number ${__data.fields.numberField2} string ${__data.fields.stringField2}"
                  }
                ]
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 17,
        "w": 16,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "maxHeight": 600,
          "mode": "multi",
          "sort": "none"
        }
      },
      "targets": [
        {
          "csvContent": "time, numberField, stringField\n1667809287000,1,test1\n1667812887000,2,test2\n1667816487000,3,test3",
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        },
        {
          "csvContent": "time, numberField2, stringField2\n1667811087000,4,test4\n1667814687000,5,test5",
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "B",
          "scenarioId": "csv_content"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "2022-11-07T08:21:27.000Z",
    "to": "2022-11-07T10:21:27.000Z"
  },
  "timeRangeUpdatedDuringEditOrView": false,
  "timepicker": {},
  "timezone": "browser",
  "title": "data-links",
  "uid": "dde7asgxmt9mof",
  "version": 7,
  "weekStart": ""
}
```
</details>

image

### CODE REVIEW

File: public/app/core/components/GraphNG/GraphNG.tsx [hunk 1]
```diff
@@ -105,30 +110,68 @@ export class GraphNG extends Component<GraphNGProps, GraphNGState> {
...
+
+      if (withLinks) {
+        const timeZone = Array.isArray(this.props.timeZone) ? this.props.timeZone[0] : this.props.timeZone;
+
+        alignedFrame.fields.forEach((field) => {
+          field.getLinks = getLinksSupplier(
+            alignedFrame,
+            field,
+            {
+              //...field.state?.scopedVars,
```
Comment (leeoniya) [inline public/app/core/components/GraphNG/GraphNG.tsx:142]: is this necessary or no? :shrug:

**Review by gillesdemey — COMMENTED** [review]
Alerting changes LGTM but will defer approval to other teams :)

**Comment by baldm0mma** [comment]
Is it an issue that the linked dashboard above's datalinks don't resolve?
Screenshot 2024-02-29 at 8 39 57 AM

**Review by baldm0mma — APPROVED** [review]
Datalinks in all my locally-tested cases seem to work as expected. Nice work!

**Pull request merged**